torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -19,13 +21,56 @@ class FSDPState(_State):
     def __init__(self):
         super().__init__()
         self._fsdp_param_group: Optional[FSDPParamGroup] = None
+        self._is_root: Optional[bool] = None
+        self._training_state: TrainingState = TrainingState.IDLE
+
+        # Attributes only used on the root state:
```

Comment:
did you consider making a subclass (i.e. RootFSDPState) to encapsulate all additional logics, instead of unifying them in the FSDPState? Is it because the `fully_shard` only accept a single state class?

Reply:
Yes, we just use one `FSDPState` class since `fully_shard`'s `contract` only accepts one class and the root is not known until 1st forward.

By the way, I refactored this part a bit in one of the later PRs recently (https://github.com/pytorch/pytorch/pull/118004).